{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9df60700",
   "metadata": {},
   "source": [
    "### \"\"\"Upload the nucleotide sequence\n",
    "    \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b0e96dfa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The DNA nucleotide sequence is:\n",
      "ATGC\n",
      "The converted RNA sequence is: AUGC\n"
     ]
    }
   ],
   "source": [
    "DNA_Seq = \"AtgC\".upper()\n",
    "RNA_Seq = \"\"\n",
    "\n",
    "print(\"The DNA nucleotide sequence is:\")\n",
    "for i in range(0, len(DNA_Seq)):\n",
    "    \n",
    "    print(DNA_Seq[i], end = \"\")\n",
    "    if DNA_Seq[i] == \"T\":\n",
    "        RNA_Seq = RNA_Seq + \"U\"\n",
    "    else :\n",
    "        RNA_Seq = RNA_Seq + DNA_Seq[i]\n",
    "print(\"\\nThe converted RNA sequence is:\", RNA_Seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "11d92afb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter a sequence:  atgggc\n",
      "\u001b[1;32mThe entered DNA sequence is legit.\n",
      "Here comes your\n",
      "\u001b[1;32mRNA sequence : \n",
      "AUGGGC\n"
     ]
    }
   ],
   "source": [
    "DNA_Seq=input(\"Enter a sequence:  \").upper()\n",
    "RNA_Seq = \"\"\n",
    "m = 0\n",
    "check_seq = False\n",
    "\n",
    "while not check_seq:\n",
    "    check_seq = True\n",
    "    for i in range(len(DNA_Seq)):\n",
    "        if DNA_Seq[i] not in \"ATGC\":\n",
    "            check_seq = False\n",
    "    if not check_seq:\n",
    "        print(\"\\u001b[31mYou have given Invalid DNA sequence\")\n",
    "        exit()\n",
    "print(\"\\033[1;32mThe entered DNA sequence is legit.\\nHere comes your\")\n",
    "\n",
    "\n",
    "print(\"\\033[1;32mRNA sequence : \")\n",
    "while m < len(DNA_Seq):\n",
    "        \n",
    "    if DNA_Seq [m] == \"A\" or DNA_Seq [m] == \"G\" or DNA_Seq [m] == \"C\":\n",
    "        RNA_Seq = RNA_Seq + DNA_Seq[m]\n",
    "        m += 1\n",
    "        \n",
    "    elif DNA_Seq[m] == \"T\":\n",
    "        RNA_Seq = RNA_Seq + \"U\"\n",
    "        m += 1\n",
    "        \n",
    "print( RNA_Seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f7a563c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCAT\n",
      "The complement of the Dna is: CGTA\n",
      "The reverse complement of the DNA is: ATGC \n"
     ]
    }
   ],
   "source": [
    "DNA_Seq = \"GCAT\".upper()\n",
    "Complement = \"\"\n",
    "Rev_Complement = \" \"\n",
    "\n",
    "for i in range (0, len(DNA_Seq)):\n",
    "    print(DNA_Seq[i], end = \"\")\n",
    "    \n",
    "    if DNA_Seq[i] == \"A\":\n",
    "        Complement = Complement +\"T\" \n",
    "        Rev_Complement = \"T\" + Rev_Complement   \n",
    "    elif DNA_Seq[i] == \"T\":\n",
    "        Complement = Complement +\"A\" \n",
    "        Rev_Complement = \"A\" + Rev_Complement        \n",
    "    elif DNA_Seq[i] == \"G\":\n",
    "        Complement = Complement +\"C\"\n",
    "        Rev_Complement = \"C\" + Rev_Complement        \n",
    "    elif DNA_Seq[i] == \"C\":\n",
    "        Complement = Complement +\"G\" \n",
    "        Rev_Complement = \"G\" + Rev_Complement        \n",
    "    else :\n",
    "        print(\"Invalid syntax\")\n",
    "print(\"\\nThe complement of the Dna is:\", Complement)\n",
    "print(\"The reverse complement of the DNA is:\", Rev_Complement)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c9843aa3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter a sequence:  sssssttywgaaliii\n",
      "Your entered sequence is:  SSSSSTTYWGAALIII\n",
      "\u001b[1;32mThe entered Protein sequence is legit.\n",
      "The total length of your Protein sequence is:  16 \n",
      "Here comes your Amino acid count: \n",
      "\n",
      "The composition for each amino acid is:\n",
      "Alanine:  12.5 % \n",
      "Cysteine:  0.0 % \n",
      "Aspartate:  0.0 % \n",
      "Glutamate:  0.0 % \n",
      "Phenylalanine:  0.0 %\n",
      "Glycine:  6.25 % \n",
      "Histidine:  0.0 % \n",
      "Isoleucine:  18.75 % \n",
      "Lysine:  0.0 % \n",
      "Leucine: 6.25 %\n",
      "Methionine:  0.0 % \n",
      "Aspargine:  0.0 % \n",
      "Proline:  0.0 % \n",
      "Glutamine:  0.0 % \n",
      "Arginine:  0.0 %\n",
      "Serine:  31.25 % \n",
      "Threonine:  12.5 % \n",
      "Valine:  0.0 % \n",
      "Tryptophan:  6.25 % \n",
      "Tyrosine is 6.25 %\n"
     ]
    }
   ],
   "source": [
    "Protein_Sequence = input(\"Enter a sequence:  \").upper()\n",
    "Pro_len = len(Protein_Sequence)\n",
    "print(\"Your entered sequence is: \", Protein_Sequence)\n",
    "m = 0\n",
    "check_seq = False\n",
    "while not check_seq:\n",
    "    check_seq = True\n",
    "    for i in range(Pro_len):\n",
    "        if Protein_Sequence[i] not in \"ACDEFGHIKLMNPQRSTVWY\":\n",
    "            check_seq = False\n",
    "    if not check_seq:\n",
    "        print(\"\\u001b[31mYou have given Invalid Protein sequence\")\n",
    "        exit()\n",
    "\n",
    "\n",
    "A_comp = (Protein_Sequence.count(\"A\") / Pro_len) * 100\n",
    "C_comp = (Protein_Sequence.count(\"C\") / Pro_len) * 100\n",
    "D_comp = (Protein_Sequence.count(\"D\") / Pro_len) * 100\n",
    "E_comp = (Protein_Sequence.count(\"E\") / Pro_len) * 100\n",
    "F_comp = (Protein_Sequence.count(\"F\") / Pro_len) * 100\n",
    "\n",
    "G_comp = (Protein_Sequence.count(\"G\") / Pro_len) * 100\n",
    "H_comp = (Protein_Sequence.count(\"H\") / Pro_len) * 100\n",
    "I_comp = (Protein_Sequence.count(\"I\") / Pro_len) * 100\n",
    "K_comp = (Protein_Sequence.count(\"K\") / Pro_len) * 100\n",
    "L_comp = (Protein_Sequence.count(\"L\") / Pro_len) * 100\n",
    "\n",
    "M_comp = (Protein_Sequence.count(\"M\") / Pro_len) * 100\n",
    "N_comp = (Protein_Sequence.count(\"N\") / Pro_len) * 100\n",
    "P_comp = (Protein_Sequence.count(\"P\") / Pro_len) * 100\n",
    "Q_comp = (Protein_Sequence.count(\"Q\") / Pro_len) * 100\n",
    "R_comp = (Protein_Sequence.count(\"R\") / Pro_len) * 100\n",
    "\n",
    "S_comp = (Protein_Sequence.count(\"S\") / Pro_len) * 100\n",
    "T_comp = (Protein_Sequence.count(\"T\") / Pro_len) * 100\n",
    "V_comp = (Protein_Sequence.count(\"V\") / Pro_len) * 100\n",
    "W_comp = (Protein_Sequence.count(\"W\") / Pro_len) * 100\n",
    "Y_comp = (Protein_Sequence.count(\"Y\") / Pro_len) * 100\n",
    "\n",
    "print(\"\\033[1;32mThe entered Protein sequence is legit.\\nThe total length of your Protein sequence is: \", len(Protein_Sequence), \"\\nHere comes your Amino acid count: \\n\")\n",
    "print(\"The composition for each amino acid is:\")\n",
    "\n",
    "print(\"Alanine: \", A_comp, \"%\", \"\\nCysteine: \", C_comp, \"%\", \"\\nAspartate: \", D_comp, \"%\", \"\\nGlutamate: \", E_comp, \"%\", \"\\nPhenylalanine: \", F_comp, \"%\")\n",
    "print(\"Glycine: \", G_comp, \"%\", \"\\nHistidine: \", H_comp, \"%\", \"\\nIsoleucine: \", I_comp, \"%\", \"\\nLysine: \", K_comp, \"%\", \"\\nLeucine:\", L_comp, \"%\")\n",
    "print(\"Methionine: \", M_comp, \"%\", \"\\nAspargine: \", N_comp, \"%\", \"\\nProline: \", P_comp, \"%\", \"\\nGlutamine: \", Q_comp, \"%\", \"\\nArginine: \", R_comp, \"%\" )\n",
    "print(\"Serine: \", S_comp, \"%\", \"\\nThreonine: \", T_comp, \"%\", \"\\nValine: \", V_comp, \"%\", \"\\nTryptophan: \", W_comp, \"%\", \"\\nTyrosine is\", Y_comp, \"%\" )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0c395b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "pro_table = {\n",
    "    \"E\": 1.51, \"M\": 1.45, \"A\": 1.42, \"L\": 1.21, \"K\": 1.16, \"F\": 1.13, \"Q\": 1.11, \"W\": 1.08,\n",
    "    \"I\": 1.08, \"V\": 1.06, \"D\": 1.01, \"H\": 1.00, \"R\": 0.96, \"T\": 0.83, \"S\": 0.77, \"C\": 0.70,\n",
    "    \"Y\": 0.69, \"N\": 0.67, \"P\": 0.57, \"G\": 0.57\n",
    "}\n",
    "\n",
    "def alpha_helix(sequence, position):\n",
    "    # TASK 1: SLICING SEQ INTO FRAMES OF SIX\n",
    "    frame1 = sequence[position:position + 6]\n",
    "    print(\"FRAME1 =\", frame1)\n",
    "\n",
    "    # TASK 2: REFER TO DICT AND CHECK PROPENSITY\n",
    "    frame2 = \"\"\n",
    "    for x in frame1:\n",
    "        if pro_table[x] >= 1:\n",
    "            frame2 += x \n",
    "    print(\"FRAME2 =\", frame2)\n",
    "\n",
    "    # CALCULATE AVG PROPENSITY IF 4 AMINO ACIDS ARE IN FRAME 2\n",
    "    if len(frame2) >= 4:\n",
    "        avg_propensity = sum(pro_table[x] for x in frame1) / len(frame1)\n",
    "        print(\"This is a alpha helix nucleation site with propensity \", avg_propensity , \"at position\", position)\n",
    "\n",
    "seq = input(\"Enter a protein sequence: \")\n",
    "position = 0\n",
    "for x in seq:\n",
    "    while position<len(seq):\n",
    "        alpha_helix(seq,position)\n",
    "        position+=6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "997dad15",
   "metadata": {},
   "outputs": [],
   "source": [
    "dnaseq = input(\"Please enter a DNA sequence.\")\n",
    "rna_seq=\" \"  #to store RNA sequence\n",
    "char_to_convert='atgc'\n",
    "STATE=1\n",
    "refrence=\"AUG\"\n",
    "i=0\n",
    "protein_list=\" \"\n",
    "Codons=\" \"\n",
    "\n",
    "codontab = {\n",
    "    'UCA': 'S', 'UCC': 'S', 'UCG': 'S', 'UCU': 'S', 'UUC': 'F',    \n",
    "    'UUU': 'F', 'UUA': 'L', 'UUG': 'L', 'UAC': 'Y', 'UAU': 'Y', 'UGC': 'C', 'UGU': 'C', 'UGG': 'W',    \n",
    "    'CUA': 'L', 'CUC': 'L', 'CUG': 'L', 'CUU': 'L', 'CCA': 'P', 'CCC': 'P', 'CCG': 'P', 'CCU': 'P',    \n",
    "    'CAC': 'H', 'CAU': 'H', 'CAA': 'Q', 'CAG': 'Q', 'CGA': 'R', 'CGC': 'R', 'CGG': 'R', 'CGU': 'R',    \n",
    "    'AUA': 'I', 'AUC': 'I', 'AUU': 'I', 'AUG': 'M', 'ACA': 'T', 'ACC': 'T', 'ACG': 'T', 'ACU': 'T',    \n",
    "    'AAC': 'N', 'AAU': 'N', 'AAA': 'K', 'AAG': 'K', 'AGC': 'S', 'AGU': 'S', 'AGA': 'R', 'AGG': 'R',    \n",
    "    'GUA': 'V', 'GUC': 'V', 'GUG': 'V', 'GUU': 'V', 'GCA': 'A', 'GCC': 'A', 'GCG': 'A', 'GCU': 'A',    \n",
    "    'GAC': 'D', 'GAU': 'D', 'GAA': 'E', 'GAG': 'E', 'GGA': 'G', 'GGC': 'G', 'GGG': 'G', 'GGU': 'G'     \n",
    "}\n",
    "#validating the converted sequence \n",
    "for x in dnaseq:\n",
    "    if (x=='A'or x=='G'or x=='T'or x=='C'):\n",
    "            if x=='T':\n",
    "                rna_seq=rna_seq+'U'\n",
    "            else:\n",
    "                rna_seq=rna_seq+x        \n",
    "    else:\n",
    "        STATE=0\n",
    "        print(\"the dna seq is invalid \",x)\n",
    "        break \n",
    "        \n",
    "if STATE==1:\n",
    "    print(\"the dna seq is \\n\"+converted_dna_seq)\n",
    "    print(\"the rna seq is \\n\" + rna_seq)\n",
    "\n",
    "while i<len(rna_seq):\n",
    "    if rna_seq[i:i+len(refrence)]==refrence:\n",
    "        try:\n",
    "            for i in range(0, len(rna_seq), 3):\n",
    "                Codons = (rna_seq[i], rna_seq[i+1], rna_seq[i+2])\n",
    "                Codons = ''.join(Codons)  # Concatenate the codons\n",
    "                print(Codons)\n",
    "                protein_list=protein_list+(codontab[Codons])\n",
    "                                \n",
    "        except IndexError:\n",
    "            print(\"Cannot read in frames of 3 anymore.\")\n",
    "            \n",
    "        finally:\n",
    "                print(\"THE PROTEIN SEQ IS\",protein_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be358b66",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fasta_reader(file_path):\n",
    "    sequences = {}\n",
    "    with open(file_path, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "        key = None\n",
    "        value = ''\n",
    "        for line in lines:\n",
    "            line = line.rstrip()  # Remove trailing whitespace characters\n",
    "            if line.startswith('>'):\n",
    "                if key is not None:\n",
    "                    sequences[key] = value\n",
    "                key = line\n",
    "                value = ''\n",
    "            else:\n",
    "                value += line\n",
    "        if key is not None:\n",
    "            sequences[key] = value\n",
    "    return sequences\n",
    "\n",
    "file_path = \"C:\\\\Users\\\\jeyar\\\\Downloads\\\\Covid-spike.txt\"\n",
    "output_dict = fasta_reader(file_path)\n",
    "#print(output_dict)\n",
    "#To make the dictionary more readable\n",
    "for key, value in output_dict.items():\n",
    "    print(f\"{key}\\n{value}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38fed16e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
